src/actions-util.ts

```diff
@@ -0,0 +1,332 @@
+  analysis_key: string;
+  // Value of the matrix for this instantiation of the job
+  matrix_vars?: string;
+  // Commit oid that the workflow was triggered on
+  commit_oid: string;
+  // Ref that the workflow was triggered on
+  ref: string;
+  // Name of the action being executed
+  action_name: ActionName;
+  // Version if the action being executed, as a commit oid
+  action_oid: string;
+  // Time the first action started. Normally the init action
+  started_at: string;
+  // Time this action started
+  action_started_at: string;
+  // Time this action completed, or undefined if not yet completed
+  completed_at?: string;
+  // State this action is currently in
+  status: ActionStatus;
+  // Cause of the failure (or undefined if status is not failure)
+  cause?: string;
+  // Stack trace of the failure (or undefined if status is not failure)
+  exception?: string;
+}
+
+/**
+ * Compose a StatusReport.
+ *
+ * @param actionName The name of the action, e.g. 'init', 'finish', 'upload-sarif'
+ * @param status The status. Must be 'success', 'failure', or 'starting'
+ * @param startedAt The time this action started executing.
+ * @param cause  Cause of failure (only supply if status is 'failure')
+ * @param exception Exception (only supply if status is 'failure')
+ */
+export async function createStatusReportBase(
+  actionName: ActionName,
+  status: ActionStatus,
+  actionStartedAt: Date,
+  cause?: string,
+  exception?: string
+): Promise<StatusReportBase> {
+  const commitOid = process.env["GITHUB_SHA"] || "";
+  const ref = getRef();
+  const workflowRunIDStr = process.env["GITHUB_RUN_ID"];
+  let workflowRunID = -1;
+  if (workflowRunIDStr) {
+    workflowRunID = parseInt(workflowRunIDStr, 10);
+  }
+  const workflowName = process.env["GITHUB_WORKFLOW"] || "";
+  const jobName = process.env["GITHUB_JOB"] || "";
+  const analysis_key = await getAnalysisKey();
+  let workflowStartedAt = process.env[sharedEnv.CODEQL_WORKFLOW_STARTED_AT];
+  if (workflowStartedAt === undefined) {
+    workflowStartedAt = actionStartedAt.toISOString();
+    core.exportVariable(
+      sharedEnv.CODEQL_WORKFLOW_STARTED_AT,
+      workflowStartedAt
+    );
+  }
+
+  const statusReport: StatusReportBase = {
+    workflow_run_id: workflowRunID,
+    workflow_name: workflowName,
+    job_name: jobName,
+    analysis_key,
+    commit_oid: commitOid,
+    ref,
+    action_name: actionName,
+    action_oid: "unknown", // TODO decide if it's possible to fill this in
+    started_at: workflowStartedAt,
+    action_started_at: actionStartedAt.toISOString(),
+    status,
+  };
+
+  // Add optional parameters
+  if (cause) {
+    statusReport.cause = cause;
+  }
+  if (exception) {
+    statusReport.exception = exception;
+  }
+  if (status === "success" || status === "failure" || status === "aborted") {
+    statusReport.completed_at = new Date().toISOString();
+  }
+  const matrix: string | undefined = core.getInput("matrix");
```

Comment:
Indeed! Code scanning is correct here!
```suggestion
  const matrix = getOptionalInput("matrix");
```

Reply:
I agree, although it looks like I got the message in the query the wrong way around. The alert is saying to use `getRequiredInput` and it should be `getOptionalInput`.

Reply:
No, the query is correct. The `matrix` input has a default value, so `getRequiredInput` is correct and the type should just be `string`.